### PR TITLE
GDV-13: [Java] Add java bindings for filter expr

### PR DESCRIPTION
**Pull request by pravindra**

### CODE REVIEW

File: java/src/main/java/org/apache/arrow/gandiva/evaluator/Filter.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+   * @param selectionVector  Result of applying the filter on the data
+   */
+  public void evaluate(ArrowRecordBatch recordBatch, SelectionVector selectionVector)
+      throws GandivaException {
+
+    if (this.closed) {
+      throw new EvaluatorClosedException();
+    }
+
+    int numRows = recordBatch.getLength();
```
Comment (praveenbingo) [inline java/src/main/java/org/apache/arrow/gandiva/evaluator/Filter.java]: should we also cross validate the record batch schema with make schema here to fail fast..this is not present in projector too..maybe add in both?
(looks like we do that in projector and filter in gandiva core..)

File: cpp/src/jni/native_builder.cc [hunk 2]
```diff
@@ -27,19 +27,22 @@
 
 #include "Types.pb.h"
```
Comment (praveenbingo) [inline cpp/src/jni/native_builder.cc]: should we rename this file too?

File: cpp/include/gandiva/selection_vector.h [hunk 3]
```diff
@@ -48,111 +48,45 @@ class SelectionVector {
...
   /// \param[in] : bitmap the bitmap
   /// \param[in] : bitmap_size size of the bitmap in bytes
   /// \param[in] : max_bitmap_index max valid index in bitmap (can be lesser than
   ///              capacity in the bitmap, due to alignment/padding).
   Status PopulateFromBitMap(const uint8_t *bitmap, int bitmap_size, int max_bitmap_index);
-};
-
-/// \brief template implementation of selection vector with a specific ctype and arrow
-/// type.
-template <typename C_TYPE, typename A_TYPE>
```
Comment (praveenbingo) [inline cpp/include/gandiva/selection_vector.h]: nice one to move this internal :)

File: cpp/src/jni/native_builder.cc [hunk 4]
```diff
@@ -527,7 +508,7 @@ Java_org_apache_arrow_gandiva_evaluator_NativeBuilder_buildNativeCode(
   // store the result in a map
   holder = std::shared_ptr<ProjectorHolder>(
       new ProjectorHolder(schema_ptr, ret_types, std::move(projector)));
-  module_id = MapInsert(holder);
+  module_id = projector_modules_.Insert(holder);
   releaseInput(schema_arr, schema_bytes, exprs_arr, exprs_bytes, env);
```
[the remark this answers is not in the document]
Comment (pravindra) [reply cpp/src/jni/native_builder.cc]: will do,

File: cpp/src/jni/native_builder.cc [hunk 5]
```diff
@@ -423,6 +395,16 @@ ExpressionPtr ProtoTypeToExpression(const types::ExpressionRoot &root) {
   return TreeExprBuilder::MakeExpression(root_node, field);
 }
 
+ConditionPtr ProtoTypeToCondition(const types::Condition &condition) {
+  NodePtr root_node = ProtoTypeToNode(condition.root());
+  if (root_node == nullptr) {
+    std::cerr << "Unable to create expression node from condition protobuf\n";
```
[the remark this answers is not in the document]
Comment (pravindra) [reply cpp/src/jni/native_builder.cc]: will remove duplicate msg

File: cpp/src/jni/native_builder.cc [hunk 6]
```diff
@@ -663,7 +645,195 @@ JNIEXPORT void JNICALL Java_org_apache_arrow_gandiva_evaluator_NativeBuilder_eva
...
+
+  // store the result in a map
+  holder = std::shared_ptr<FilterHolder>(new FilterHolder(schema_ptr, std::move(filter)));
+  module_id = filter_modules_.Insert(holder);
+  releaseFilterInput(schema_arr, schema_bytes, condition_arr, condition_bytes, env);
+  return module_id;
+
+err_out:
+  env->ThrowNew(gandiva_exception_, ss.str().c_str());
+  return module_id;
```
[the remark this answers is not in the document]
Comment (pravindra) [reply cpp/src/jni/native_builder.cc]: both project/filter return 0 on error. module_id can't be 0.

File: cpp/src/jni/native_builder.cc [hunk 7]
```diff
@@ -663,7 +645,195 @@ JNIEXPORT void JNICALL Java_org_apache_arrow_gandiva_evaluator_NativeBuilder_eva
...
+  std::shared_ptr<gandiva::SelectionVector> selection_vector;
+
+  do {
+    auto schema = holder->schema();
+    std::vector<std::shared_ptr<arrow::ArrayData>> columns;
+    auto num_fields = schema->num_fields();
+    int buf_idx = 0;
+    int sz_idx = 0;
+
+    for (int i = 0; i < num_fields; i++) {
```
[the remark this answers is not in the document]
Comment (pravindra) [reply cpp/src/jni/native_builder.cc]: will do.

File: cpp/src/jni/native_builder.cc [hunk 8]
```diff
@@ -663,7 +645,195 @@ JNIEXPORT void JNICALL Java_org_apache_arrow_gandiva_evaluator_NativeBuilder_eva
...
+        auto offsets = std::shared_ptr<arrow::Buffer>(
+            new arrow::Buffer(reinterpret_cast<uint8_t *>(offsets_addr), offsets_size));
+        buffers.push_back(offsets);
+      }
+
+      auto array_data =
+          arrow::ArrayData::Make(field->type(), num_rows, std::move(buffers));
+      columns.push_back(array_data);
+    }
+    if (!status.ok()) {
```
[the remark this answers is not in the document]
Comment (pravindra) [reply cpp/src/jni/native_builder.cc]: the macro CHECK_IN_BUFFER_IDX_AND_BREAK can modify status

File: java/src/main/java/org/apache/arrow/gandiva/evaluator/Filter.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    if (selectionVector.getMaxRecords() < numRows) {
+      logger.info("selectionVector has capacity for " + numRows
```
[the remark this answers is not in the document]
Comment (pravindra) [reply java/src/main/java/org/apache/arrow/gandiva/evaluator/Filter.java]: will switch to error

File: java/src/main/java/org/apache/arrow/gandiva/evaluator/Filter.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+/**
+ * This class provides a mechanism to filter a RecordBatch by evaluating a condition expression.
+ * Follow these steps to use this class:
+ * 1) Use the static method make() to create an instance of this class that evaluates a
+ * condition.
+ * 2) Invoke the method evaluate() to evaluate the filter against a RecordBatch
+ * 3) Invoke close() to release resources
+ */
+public class Filter {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(Filter.class);
```
[the remark this answers is not in the document]
Comment (pravindra) [reply java/src/main/java/org/apache/arrow/gandiva/evaluator/Filter.java]: will fix

File: java/src/main/java/org/apache/arrow/gandiva/evaluator/Filter.java (hunk 1 above)
Comment (pravindra) [reply java/src/main/java/org/apache/arrow/gandiva/evaluator/Filter.java]: evaluate codepath is very performance sensitive. can't afford to do the same check twice.

File: cpp/src/jni/native_builder.cc (hunk 2 above)
Comment (pravindra) [reply cpp/src/jni/native_builder.cc]: renaming to jni_common.cc

File: cpp/include/gandiva/selection_vector.h (hunk 3 above)
Comment (pravindra) [reply cpp/include/gandiva/selection_vector.h]: thanks !

**Pull request merged**